test/cpp/api/tensor.cpp

```diff
@@ -406,3 +406,31 @@ TEST(TensorTest, Requires_grad_) {
   ASSERT_THROWS_WITH(y.requires_grad_(false),
     "requires_grad_ is not implemented for Tensor");
 }
+
+TEST(TensorTest, Detach) {
+  auto x = torch::tensor({5}, at::TensorOptions().requires_grad(true));
+  auto y = x * x;
+  const auto y_detached = y.detach();
+  ASSERT_FALSE(y.is_leaf());
+  ASSERT_TRUE(y_detached.is_leaf());
+
+  x = at::tensor({5}, at::TensorOptions().requires_grad(false));
+  y = x * x;
+  const auto message = "detach is not implemented for Tensor";
+  ASSERT_THROWS_WITH(x.detach(), message);
+  ASSERT_THROWS_WITH(y.detach(), message);
+}
+
+TEST(TensorTest, Detach_) {
```

Comment:
I feel that we might want to name this `DetachInplace`, to preserve the naming convention (of not using underscore). Same for other tests as well.